### ISSUE TITLE
os_server module problem with floating_ips and multiple subnets

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29594]
From @lukasged on 2016-10-10T12:49:26Z
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

os_server
##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

host_key_checking = False
force_color = 1
scp_if_ssh = True

and also set the vault_password_file and inventory args
##### OS / ENVIRONMENT

I'm both running and managing a RedHat 7.2
##### SUMMARY

I'm creating instances on OpenStack with the ansible module os_server. I have the floating ips already assigned before executing ansible, so I want to use the floating_ips argument, so the module would assign the floating ips automatically. I have 2 networks in my tenant, each one with its subnet, and the floating ip's have to be assigned to the correct subnet. The problem is that with the floating_ips argument, I can't establish the subnet that the floating ip should be assigned to. It only works if I specify 1 floating IP. It gets correctly assigned, but then I have to assign the second one manually, or using the 'os_floating_ip' module. When I specify both floating ip's in the argument, I get the following message : 

```
failed: [localhost] (item=1) => {"extra_data": null, "failed": true, "item": "1", "msg": "unable to bind a floating ip to server *******: External network ********* is not reachable from subnet ********.  Therefore, cannot associate Port ************ with a Floating IP.\nNeutron server returns request_ids: ['req-************']"}
```
##### STEPS TO REPRODUCE

You have to be in a tenant where there are more than one subnet defined and then use the os_server module, defining the floating_ips argument with a floating IP for each subnet.
##### EXPECTED RESULTS

I expect the floating ip's to be assigned to the correct subnets
##### ACTUAL RESULTS

```
failed: [localhost] (item=1) => {"extra_data": null, "failed": true, "item": "1", "msg": "unable to bind a floating ip to server *******: External network ********* is not reachable from subnet ********.  Therefore, cannot associate Port ************ with a Floating IP.\nNeutron server returns request_ids: ['req-************']"}
```

Copied from original issue: ansible/ansible-modules-core#5209

**Comment by ansibot** [comment]
From @ansibot on 2016-10-10T12:49:26Z



@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @geektoor on 2016-10-10T12:49:26Z

Hey,

i ran into the same issue. After i dug a bit into it, i found a "side note" here:
http://docs.openstack.org/infra/shade/usage.html
at add_auto_ip:
> For advanced network architecture (e.g. multiple external networks or servers with multiple interfaces), use other floating IP methods.

I think this is a "bit" bad, as auto_ip is set to yes by default. And even if i specify a floating ip pool, it fails, because it tries to assign an IP from the first floating pool in the list.
Suggestion: make at least ip_pool working when auto_ip defaults to yes, otherwise change the default to no as it might not work in any advanced setup (multiple floating pools, multiple network interfaces...).

Cheers,
Sven

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (122/100 (122%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext8.iad.github.net ([192.30.252.199]:42357 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ2g-000oSq-Bx
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:50:30 -0500
Date: Mon, 11 Sep 2017 07:49:21 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141382;
	bh=CxBx04Y46Hemkhm3vWb8XtuRktNPk2cX3Cjlh+SBDzY=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=xSL2ETDXUUgvgT1ZF4yeQwCp89dABQFlI/xDuSUVsBF2R+4cisr1mHS7UrRdR0wJ9
	 r+B5Qs30jeu6LGKv4JIpTehpnvTRxevxlFe+ngYdrhULZ04OaKrtzGMYSVlJjwqGBo
	 nLnXXXjhIANonfL+0xoxXJlXImhtNdt2waNSuE6o=

**Comment by ansibot** [comment]
cc @dagnello @thingee
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @omgjlk
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @mnaser
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mnaser** [comment]
wontfix

This is a pretty stale and old bug.  You should be able to assign floating IPs using https://docs.ansible.com/ansible/2.6/modules/os_floating_ip_module.html to the exact port/floating ip/subnet you need.

**Comment by panticz** [comment]
Since Ansible 2.3. you can define nat_destination: YOUR_NETWORK in the os_floating_ip module:

- os_floating_ip:
    ...
     nat_destination: private_net
